### PR TITLE
Add Kerberos authentication support

### PR DESCRIPTION
**Pull request by ggreg**

### CODE REVIEW

**Comment by ggreg** [comment]
To test this on a specific server, set the following environment variables:
- `PRESTO_RUNNING_HOST`
- `PRESTO_RUNNING_PORT`

You can also set `SOCKS_PROXY` to a `HOST:PORT` server that will be use to proxy HTTP requests of Presto queries.

**Comment by highker** [comment]
I will make a release with this patch (https://github.com/highker/presto-python-client/releases); let's merge this when the tests are complete

**Comment by highker** [comment]
two other things it would nice to have in this patch are
- retry when auth fails (this can be patched together with https://github.com/prestodb/presto-python-client/issues/31)
- may worth introducing a new error type say `PrestoAuthenticationError`

**Comment by ggreg** [comment]
- Thanks for PR #33!
- I thought about defining a simple protocol to have authentication as a *middleware*, so we could compose multiple handler that act on the HTTP requests and Presto client session, but I think this can wait until we have another candidate in addition to authentication.

**Comment by ggreg** [comment]
@highker any reason to use `requests_kerberos==0.10.0` rather than `0.11.0`, the latest?

File: prestodb/auth.py [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        self._ca_bundle = ca_bundle
+
+    def set_client_session(self, client_session):
+        pass
+
+    def set_http_session(self, http_session):
+        if self._config:
+            os.environ['KRB5_CONFIG'] = self._config
+        http_session.trust_env = False
+        http_session.auth = requests_kerberos.HTTPKerberosAuth(
```
Comment (ggreg) [inline prestodb/auth.py]: I ended up exposing all arguments of [`HTTPKerberosAuth`'s constructor](https://github.com/requests/requests-kerberos/blob/master/requests_kerberos/kerberos_.py#L85).

File: prestodb/client.py [hunk 2]
```diff
@@ -301,14 +304,34 @@ def next_uri(self):
...
+        http_response = self._post(
             self.statement_url,
-            data=sql.encode('utf-8'),
-            headers=self.http_headers,
+            data=data,
+            headers=http_headers,
             timeout=self._request_timeout,
             proxies=PROXIES,
         )
+        if allow_redirect and self._redirect_handler is not None:
```
Comment (ggreg) [inline prestodb/client.py]: @highker the redirect logic is called here.

File: prestodb/redirect.py [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import abc
+from future.standard_library import install_aliases
+install_aliases()
+
+from future.utils import with_metaclass
+import ipaddress
```
Comment (highker) [inline prestodb/redirect.py]: can we add this to setup.py?

File: tests/test_client.py [hunk 4]
```diff
@@ -544,3 +545,42 @@ def test_503_error_retry(monkeypatch):
 
     req.get('URL')
     assert post_retry.retry_count == attempts
+
+
+class FakeGatewayResponse(object):
+    def __init__(self, http_response, nb_redirects=1):
+        self.http_response = http_response
```
Comment (highker) [inline tests/test_client.py]: let's add `self.__name__ = 'FakeGatewayResponse'`

File: prestodb/client.py [hunk 5]
```diff
@@ -301,14 +304,34 @@ def next_uri(self):
         # type: () -> Text
         return self._next_uri
 
-    def post(self, sql):
-        return self._post(
+    def post(self, sql, allow_redirect=False):
```
Comment (highker) [inline prestodb/client.py]: maybe `self._redirect_handler is not None` is good enough; users do not have ability to make this `True`

File: prestodb/client.py [hunk 6]
```diff
@@ -301,14 +304,34 @@ def next_uri(self):
...
+        data = sql.encode('utf-8')
+        http_headers = self.http_headers
+
+        http_response = self._post(
```
Comment (highker) [inline prestodb/client.py]: need to toggle `allow_redirects=False`

**Comment by highker** [comment]
@ggreg, Is this patch ready for review? I chatted with few people in other teams; kerberos authentication is not easy to test with. So I assume we could leave the component as is.

**Comment by ggreg** [comment]
Let's merge it for some internal testing. We will add proper integration tests than run on Travis in follow-up tasks.

File: setup.py [hunk 7]
```diff
@@ -58,6 +58,7 @@
     install_requires=[
         'click==6.7',
         'future==0.16.0',
+        'ipaddress==1.0.18',
```
Comment (highker) [inline setup.py]: For the same reason as `requests_kerberos `, can we change it to 1.0.14?

File: prestodb/client.py [hunk 8]
```diff
@@ -302,13 +305,34 @@ def next_uri(self):
...
-            data=sql.encode('utf-8'),
-            headers=self.http_headers,
+            data=data,
+            headers=http_headers,
             timeout=self._request_timeout,
+            allow_redirects=False,
             proxies=PROXIES,
         )
+        if self._redirect_handler is not None:
+            while http_response.is_redirect:
```
Comment (highker) [inline prestodb/client.py]: `while http_response is not None and http_response.is_redirect:`

File: prestodb/client.py [hunk 9]
```diff
@@ -302,13 +305,34 @@ def next_uri(self):
...
+        http_headers = self.http_headers
+
+        http_response = self._post(
             self.statement_url,
-            data=sql.encode('utf-8'),
-            headers=self.http_headers,
+            data=data,
+            headers=http_headers,
             timeout=self._request_timeout,
+            allow_redirects=False,
```
Comment (highker) [inline prestodb/client.py]: I think this should be `allow_redirects=self._redirect_handler is None`

File: prestodb/client.py [hunk 10]
```diff
@@ -302,13 +305,34 @@ def next_uri(self):
...
+                    http_response.status_code,
+                    location,
+                    url,
+                ))
+                http_response = self._post(
+                    url,
+                    data=data,
+                    headers=http_headers,
+                    timeout=self._request_timeout,
+                    proxies=PROXIES,
```
Comment (highker) [inline prestodb/client.py]: Add ``allow_redirects=False`

File: prestodb/redirect.py [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+import socket
+from typing import Any  # NOQA
+from urllib.parse import urlparse
+
+
+class RedirectHandler(with_metaclass(abc.ABCMeta)):
+    @abc.abstractmethod
+    def handle(self, url):
+        pass
+
```
Comment (highker) [inline prestodb/redirect.py]: remove extra line

**Comment by ggreg** [comment]
@highker thanks for the review. I made the suggested changes. Assigning back to you if you have still comments. Otherwise assign it back to me and I will merge this PR.

File: prestodb/redirect.py [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+
+
+def _normalize_url_with_hostname(url):
+    parsed = urlparse(url.encode('utf-8'))
```
Comment (highker) [inline prestodb/redirect.py]: Let's add a TODO here; `urllib.urlunparse` is bogus dealing with utf8; there may have a potential problem with `urlparse` too; replace the library in the future if possible.

File: tests/test_client.py [hunk 13]
```diff
@@ -544,3 +546,43 @@ def test_503_error_retry(monkeypatch):
 
     req.get('URL')
     assert post_retry.retry_count == attempts
+
+
+class FakeGatewayResponse(object):
+    def __init__(self, http_response, nb_redirects=1):
```
Comment (highker) [inline tests/test_client.py]: `nb_redirects` --> `redirect_count`

**Comment by jeffreythewang** [comment]
Hey @ggreg, I was trying to set up a Kerberos-authenticated connection from a python client to a presto instance using this module, and I was wondering how you tested this. Were you connecting to presto via SSL over HTTPS? Did you set `ca_bundle`, and if so what kind of certificate did you use? Were there any keytab files involved?

I have a similar set up working successfully with the [JDBC client](https://prestodb.io/docs/current/installation/jdbc.html), and I wanted to translate those configurations over to a python client, if possible.

Thanks

**Comment by ggreg** [comment]
@jeffreythewang an example is:

```python
import requests_kerberos

KRB5_CONFIG = "/etc/krb5-thrift.conf"
CA_BUNDLE = "/etc/pki/tls/certs/certs.pem"

_auth = KerberosAuthentication(
    config=KRB5_CONFIG,
    service_name=COORDINATOR_KRB5_SERVICE_NAME,
    mutual_authentication=requests_kerberos.DISABLED,
    ca_bundle=CA_BUNDLE,
)
```

**Pull request merged**